src/posix/platform/radio_spinel.cpp

```diff
@@ -601,27 +601,45 @@ void RadioSpinel::HandleValueIs(spinel_prop_key_t aKey, const uint8_t *aBuffer,
 otError RadioSpinel::ParseRadioFrame(otRadioFrame &aFrame, const uint8_t *aBuffer, uint16_t aLength)
 {
     otError        error        = OT_ERROR_NONE;
-    uint16_t       packetLength = 0;
+    uint16_t       flags        = 0;
+    int8_t         noiseFloor   = -128;
+    spinel_size_t  size         = OT_RADIO_FRAME_MAX_SIZE;
+    unsigned int   receiveError = 0;
     spinel_ssize_t unpacked;
-    uint16_t       flags      = 0;
-    int8_t         noiseFloor = -128;
-    spinel_size_t  size       = OT_RADIO_FRAME_MAX_SIZE;
-
-    unpacked = spinel_datatype_unpack(aBuffer, aLength, SPINEL_DATATYPE_UINT16_S, &packetLength);
-    VerifyOrExit(unpacked > 0 && packetLength <= OT_RADIO_FRAME_MAX_SIZE, error = OT_ERROR_PARSE);
-
-    aFrame.mLength = static_cast<uint8_t>(packetLength);
-
-    unpacked = spinel_datatype_unpack_in_place(aBuffer, aLength,
-                                               SPINEL_DATATYPE_DATA_WLEN_S SPINEL_DATATYPE_INT8_S SPINEL_DATATYPE_INT8_S
-                                                   SPINEL_DATATYPE_UINT16_S SPINEL_DATATYPE_STRUCT_S( // PHY-data
-                                                       SPINEL_DATATYPE_UINT8_S     // 802.15.4 channel
-                                                           SPINEL_DATATYPE_UINT8_S // 802.15.4 LQI
-                                                       ),
-                                               aFrame.mPsdu, &size, &aFrame.mInfo.mRxInfo.mRssi, &noiseFloor, &flags,
-                                               &aFrame.mChannel, &aFrame.mInfo.mRxInfo.mLqi);
+
+    // Timestamp is ms + us.
+    unpacked = spinel_datatype_unpack_in_place(
+        aBuffer, aLength,
+        SPINEL_DATATYPE_DATA_WLEN_S                              // Frame
+                        SPINEL_DATATYPE_INT8_S                   // RSSI
+                        SPINEL_DATATYPE_INT8_S                   // Noise Floor
+                        SPINEL_DATATYPE_UINT16_S                 // Flags
+                        SPINEL_DATATYPE_STRUCT_S(                // PHY-data
+                            SPINEL_DATATYPE_UINT8_S              // 802.15.4 channel
+                                        SPINEL_DATATYPE_UINT8_S  // 802.15.4 LQI
+                                        SPINEL_DATATYPE_UINT32_S // Timestamp (ms).
+                                        SPINEL_DATATYPE_UINT16_S // Timestamp (us).
+                            ) SPINEL_DATATYPE_STRUCT_S(          // Vendor-data
+                            SPINEL_DATATYPE_UINT_PACKED_S        // Receive error
+                            ),
+        aFrame.mPsdu, &size, &aFrame.mInfo.mRxInfo.mRssi, &noiseFloor, &flags, &aFrame.mChannel,
+        &aFrame.mInfo.mRxInfo.mLqi, &aFrame.mInfo.mRxInfo.mMsec, &aFrame.mInfo.mRxInfo.mUsec, &receiveError);
+
     VerifyOrExit(unpacked > 0, error = OT_ERROR_PARSE);
 
+    if (receiveError == 0)
+    {
+        aFrame.mLength = static_cast<uint8_t>(size);
```

Comment:
`spinel_size_t` is an `unsigned int`, so it's possible that `size` might be a value greater than 255. A check should be made (and an error returned to the sender if it fails) before performing the cast to the narrower type.

Reply:
`size` is initialized to `OT_RADIO_FRAME_MAX_SIZE` which is `127`. It restricts the parse to at most fills `127` bytes into the buffer, otherwise it results in parsing error. You may check it out [here](https://github.com/openthread/openthread/blob/master/src/ncp/spinel.c#L530).